### PR TITLE
chore: add updates-svc

### LINKED ISSUE #871: `updates-svc` to provide a realtime updates endpoint for our clients
### Describe the feature

The `updates-svc` should provide a gRPC server-side stream to our clients to give them the ability to listen on our EventStoreDB stream.

We currently have event sourced and non-event sourced entites. Event sourced entites are easy to implement. Non-event sourced entities need to dispatch updates events later on. The latter is not part of this issue.

### Why is this feature needed?

To enable a certain form of collaboration, the UI must always be up to date. 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by MaxSchaefer**
Closes #871

Todo before merge
- [x] Deploy to staging

### CODE REVIEW

File: libs/common/auth/auth.go [hunk 1]
```diff
@@ -26,9 +27,11 @@ var (
 	oauthConfig             *oauth2.Config
 	Verifier                *oidc.IDTokenVerifier
 	authSetupDone           bool
+	FakeTokenValidFor       = time.Second * 30
 )
 
 type ClaimsKey struct{}
+type TokenExpires struct{}
```
Comment (FoseFx) [inline libs/common/auth/auth.go:34]: btw why are these exported? they should not be!

Comment (FoseFx) [inline libs/common/auth/auth.go:34]:
```suggestion
type TokenExpiresKey struct{}
```

File: libs/common/auth/auth.go [hunk 2]
```diff
@@ -140,49 +143,50 @@ func (t IDTokenClaims) AsExpected() error {
 // VerifyIDToken verifies the correctness of the accessToken and returns its claim.
 // The claim is checked to be as expected.
 // Service must be set up with auth!
-func VerifyIDToken(ctx context.Context, token string) (*IDTokenClaims, error) {
+func VerifyIDToken(ctx context.Context, token string) (*IDTokenClaims, *time.Time, error) {
```
Comment (FoseFx) [inline libs/common/auth/auth.go:146]: document second return parameter

File: libs/common/auth/auth.go [hunk 3]
```diff
@@ -206,6 +210,16 @@ func GetAuthClaims(ctx context.Context) (*IDTokenClaims, error) {
 	}
 }
 
+// SessionValidUntil returns time.Time when the session gets marked as expired
+func SessionValidUntil(ctx context.Context) (time.Time, error) {
+	res, ok := ctx.Value(TokenExpires{}).(*time.Time)
+	if !ok {
+		return time.Now(), status.Error(codes.Internal, "tokenExpires not in context, set up auth")
```
Comment (FoseFx) [inline libs/common/auth/auth.go:217]: this is not a valid expected state, thus warrants a panic, not an error

File: proto/services/updates_svc/v1/updates_svc.proto [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  string aggregate_id = 3;
+  string aggregate_type = 4;
+}
+
+//
+// ReceiveUpdatesRequest
+//
+
+message ReceiveUpdatesRequest {
+  optional uint64 revision = 1;
```
Comment (FoseFx) [inline proto/services/updates_svc/v1/updates_svc.proto:27]: document this key

File: proto/services/updates_svc/v1/updates_svc.proto [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+//
+// ReceiveUpdatesResponse
+//
+
+message ReceiveUpdatesResponse {
+  uint64 revision = 1;
```
Comment (FoseFx) [inline proto/services/updates_svc/v1/updates_svc.proto:35]: document these keys

Comment (FoseFx) [reply proto/services/updates_svc/v1/updates_svc.proto:35]: what makes an event an entity and what a domain event?

Comment (FoseFx) [reply proto/services/updates_svc/v1/updates_svc.proto:35]: and why would the client care?

File: services/updates-svc/internal/updates/grpc.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+	return &UpdatesGrpcService{
+		es: es,
+	}
+}
+
+func (s *UpdatesGrpcService) ReceiveUpdates(req *pb.ReceiveUpdatesRequest, stream pb.UpdatesService_ReceiveUpdatesServer) error {
+	ctx, span, log := telemetry.StartSpan(stream.Context(), "updates.ReceiveUpdates")
+	defer span.End()
+
+	tokenExpires, err := auth.SessionValidUntil(ctx)
```
Comment (FoseFx) [inline services/updates-svc/internal/updates/grpc.go:28]:
```suggestion
	tokenExpiresAt, err := auth.SessionValidUntil(ctx)
```

File: services/updates-svc/internal/updates/grpc.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+		return err
+	}
+
+	tokenExpiresIn := time.Until(tokenExpires)
+
+	ctx, cancel := context.WithCancel(ctx)
+	// Close stream after session is invalid
+	time.AfterFunc(tokenExpiresIn, func() {
+		log.Debug().Msg("closing stream, token expired")
+		cancel()
```
Comment (FoseFx) [inline services/updates-svc/internal/updates/grpc.go:39]: will these be cleaned up if the stream breaks?

else I can see the GC holdon to this function, which holds on to the cancel() functions, which holds on to the context

might cause a memory leak?

it returns something you can cancel this with, we should do that (by defer- ing)

File: services/updates-svc/internal/updates/grpc.go [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+	if err != nil {
+		return err
+	}
+	defer esStream.Close()
+
+	for {
+		esdbEvent := esStream.Recv()
+
+		select {
+		case <-stream.Context().Done():
```
Comment (FoseFx) [inline services/updates-svc/internal/updates/grpc.go:80]: isn't this the wrong context? we cancel the ctx, which is a fresh new wrapped context created above.

File: services/updates-svc/internal/updates/grpc.go (hunk 7 above)
Comment (FoseFx) [reply services/updates-svc/internal/updates/grpc.go:39]: Instead use context.WithTimeout()

File: services/updates-svc/internal/updates/grpc.go (hunk 8 above)
Comment (FoseFx) [reply services/updates-svc/internal/updates/grpc.go:80]: also add a logline here

File: services/updates-svc/internal/updates/grpc.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+			return nil
+		default:
+		}
+
+		if esdbEvent.SubscriptionDropped != nil {
+			esStream.Close()
+			log.Debug().Msg("subscription dropped, close stream")
```
Comment (FoseFx) [inline services/updates-svc/internal/updates/grpc.go:87]: does esdbEvent.SubscriptionDropped supply more information, which may be interesting to be logged?

File: services/updates-svc/internal/updates/grpc.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+			break
```
Comment (FoseFx) [inline services/updates-svc/internal/updates/grpc.go:88]:
```suggestion
			return nil
```

File: services/updates-svc/internal/updates/grpc.go [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+		res := &pb.ReceiveUpdatesResponse{
+			Revision: *esdbEvent.EventAppeared.Commit,
+			Event: &pb.ReceiveUpdatesResponse_DomainEvent{
+				DomainEvent: domainEvent,
+			},
+		}
+
+		if err := stream.Send(res); err != nil {
+			log.Error().Err(err).Msg("cannot send on stream")
```
Comment (FoseFx) [inline services/updates-svc/internal/updates/grpc.go:127]:
```suggestion
			log.Error().Err(err).Msg("cannot send update response on stream")
```

File: services/updates-svc/stories/setup_test.go [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+	zlog.Info().
+		Str("eventstoreEndpoint", eventstoreEndpoint).
+		Msg("containers are up")
+
+	// prepare env
+	hwtesting.SetCommonEnv()
+	hwtesting.SetEventstoreEnv(eventstoreEndpoint)
+
+	// TODO: spice migrations (PR #812)
```
Comment (FoseFx) [inline services/updates-svc/stories/setup_test.go:33]:
```suggestion
```

File: services/updates-svc/stories/updates_test.go [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+	assert.NoError(t, err)
+	assert.NoError(t, stream.CloseSend())
+}
+
+func TestAutoClosingWhenTokenExpiresReceiveUpdateStream(t *testing.T) {
+	ctx := context.Background()
+	updatesClient := updatesServiceClient()
+
+	auth.FakeTokenValidFor = time.Second * 3
+	upperBound := auth.FakeTokenValidFor + (time.Second * 1)
```
Comment (FoseFx) [inline services/updates-svc/stories/updates_test.go:28]: there is time.Round() and time.Truncate()

File: services/updates-svc/stories/updates_test.go [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+	lowerBound := auth.FakeTokenValidFor - (time.Second * 1)
+
+	req := &pb.ReceiveUpdatesRequest{}
+
+	timeBeforeOpeningStream := time.Now()
+	stream, err := updatesClient.ReceiveUpdates(ctx, req)
+	assert.NoError(t, err)
+
+	_, cancel := context.WithTimeout(ctx, auth.FakeTokenValidFor+(time.Second*3))
```
Comment (FoseFx) [inline services/updates-svc/stories/updates_test.go:37]: why do you create a context that you dont use at all?

Comment (FoseFx) [reply services/updates-svc/stories/updates_test.go:37]: contexts are immutable

File: services/updates-svc/stories/updates_test.go [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+		if err == io.EOF {
+			elapsedTime := time.Since(timeBeforeOpeningStream)
+			if elapsedTime > upperBound || elapsedTime < lowerBound {
+				t.Errorf("Connection timeout is out of bounds (upper bound %s, lower bound %s) with %s", upperBound, lowerBound, elapsedTime)
+				return
+			}
+			return
+		}
+		assert.NoError(t, err)
+	}
```
Comment (FoseFx) [inline services/updates-svc/stories/updates_test.go:51]: what if the stream never EOFs?

File: libs/common/auth/auth.go (hunk 3 above)
Comment (MaxSchaefer) [reply libs/common/auth/auth.go:217]: https://github.com/helpwave/services/issues/811#issuecomment-2427021965

File: services/updates-svc/stories/updates_test.go (hunk 13 above)
Comment (MaxSchaefer) [reply services/updates-svc/stories/updates_test.go:29]: Uhiii very nice!

File: services/updates-svc/stories/updates_test.go (hunk 14 above)
Comment (FoseFx) [reply services/updates-svc/stories/updates_test.go:39]: why would it? the context that is being canceled is not used, and nothing else will be triggered that would warrant a failed test

Comment (MaxSchaefer) [reply services/updates-svc/stories/updates_test.go]: Now that you mention it ... weird that I tested this behaviour before and it worked. _Fixed._

**Pull request merged**